Support Alter Table Clause For External Table

HappenLee opened the pull request:
5 alter operation be supported:

* RENAME
* ADD COLUMN
* DROP COLUMN
* MODIFY COLUMN
* REORDER COLUMN

Fix #4697 


## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [] Documentation Update (if none of the other choices apply)
- [] Code refactor (Modify the code structure, format the code, etc...)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have create an issue on (Fix #ISSUE), and have described the bug/feature there in detail
- [x] Compiling and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] If this change need a document change, I have updated the document
- [x] Any dependent changes have been merged

HappenLee committed:
Support Alter Table Clause For External Table
5 alter operation be supported:

* RENAME
* ADD COLUMN
* DROP COLUMN
* MODIFY COLUMN
* REORDER COLUMN

morningman commented on fe/fe-core/src/main/java/org/apache/doris/alter/Alter.java:
```diff
@@ -172,71 +256,18 @@ public void processAlterTable(AlterTableStmt stmt) throws UserException {
+                case MYSQL:
+                case ELASTICSEARCH:
+                    processAlterExternalTable(stmt, table, db);
+                    break;
```
use `return` instead of `break` is more clear

morningman commented on fe/fe-core/src/main/java/org/apache/doris/alter/AlterHandler.java:
```diff
@@ -380,6 +381,12 @@ public void start() {
+    /*
+     * entry function. handle alter ops for external table
+     */
+    public void process(List<AlterClause> alterClauses, Database db, Table externalTable)
```
`processExternalTable` is more clear

morningman commented on fe/fe-core/src/main/java/org/apache/doris/alter/SchemaChangeHandler.java:
```diff
@@ -132,6 +132,30 @@ private void processAddColumn(AddColumnClause alterClause, OlapTable olapTable,
+
+        List<Column> modIndexSchema = externalTable.getBaseSchema();
+        addColumnInternal(externalTable, column, columnPos, modIndexSchema, newColNameSet);
+        externalTable.setNewFullSchema(modIndexSchema);
```
Should not set new schema to `externalTable` here directly.
The other alter clause may throw exception and the entire alter operation will fail, and you have to rollback all changes to the table.

A correct way to to modify a "copied" schema of external table, and set the new schema at the end of all alter clauses.

HappenLee replied:
yes， it is a problem。I will set the new schema at the end of alter external table.

morningman commented on fe/fe-core/src/main/java/org/apache/doris/catalog/Catalog.java:
```diff
@@ -5045,6 +5046,36 @@ public void renameTable(Database db, OlapTable table, TableRenameClause tableRen
+        LOG.info("rename table[{}] to {}", tableName, newTableName);
+    }
+
+    public void reflushTable(Database db, Table table) throws DdlException {
```
This is not a good way to alter external table.
You write 2 edit logs here, which will make this operation non-atomic。

HappenLee replied:
yes，to make sure it is atomic. We may need add a new OperationType in edit log.

HappenLee committed:
fix some problem

1. copy old schema to new schema before external table schema change
2. add new operation of OP_REFRESH_EXTERNAL_TABLE_SCHEMA

HappenLee committed:
change some code

wuyunfeng commented:
This PR can also be applied with `Elasticsearch` external table. This is so good

wuyunfeng commented on fe/fe-core/src/main/java/org/apache/doris/alter/Alter.java:
```diff
@@ -377,6 +408,17 @@ private void processRename(Database db, OlapTable table, List<AlterClause> alter
         }
     }
 
+    private void processRename(Database db, Table table, List<AlterClause> alterClauses) throws DdlException {
```
```suggestion
    private void processRenameAlterOperation(Database db, Table table, List<AlterClause> alterClauses) throws DdlException {
```

HappenLee replied:
@wuyunfeng hi，yunfeng. I just want keep the function name same to the olaptable alter function name.

wuyunfeng commented on fe/fe-core/src/main/java/org/apache/doris/persist/OperationType.java:
```diff
@@ -182,4 +182,7 @@
     public static final short OP_DROP_RESOURCE = 277;
+
+    // alter external table
+    public static final short OP_REFRESH_EXTERNAL_TABLE_SCHEMA = 280;
```
just puzzled with the `refresh`?

HappenLee replied:
OK，I will change the name of this operationType

wuyunfeng commented on fe/fe-core/src/main/java/org/apache/doris/alter/SchemaChangeHandler.java:
```diff
@@ -132,6 +132,26 @@ private void processAddColumn(AddColumnClause alterClause, OlapTable olapTable,
                           indexSchemaMap, newColNameSet);
     }
 
+    private void processAddColumn(AddColumnClause alterClause, Table externalTable, List<Column> newSchema) throws DdlException {
```
```suggestion
    private void processAddCol{Alter}Operation(AddColumnClause alterClause, Table externalTable, List<Column> newSchema) throws DdlException {
```

HappenLee replied:
hi，yunfeng. I just want keep the function name same to the olaptable alter function name.

wuyunfeng replied:
That's ok

HappenLee committed:
change OP_REFRESH_EXTERNAL_TABLE_SCHEMA to OP_ALTER_EXTERNAL_TABLE_SCHEMA

HappenLee committed:
Added a forbidden of delete the last column

morningman commented on fe/fe-core/src/main/java/org/apache/doris/catalog/Database.java:
```diff
@@ -305,6 +306,26 @@ public boolean createTableWithLock(Table table, boolean isReplay, boolean setIfN
         }
     }
 
+    public void refreshTableSchemaWithLock(String tableName, List<Column> newSchema, boolean isReplay) throws DdlException{
```
the param `isReplay` seems always true.

HappenLee replied:
yes, so I will del it.

morningman commented on fe/fe-core/src/main/java/org/apache/doris/persist/RefreshExternalTableInfo.java:
```diff
@@ -0,0 +1,102 @@
+public class RefreshExternalTableInfo implements Writable {
+    public static final Logger LOG = LoggerFactory.getLogger(RefreshExternalTableInfo.class);
+
+    private String dbName;
```
Use GSON

HappenLee replied:
Sorry, Firstly I try use GSON to do serialization. But failed when I need to deserialize the `newSchema`. It seems `Column` cause the problem.

morningman commented on fe/fe-core/src/main/java/org/apache/doris/catalog/Catalog.java:
```diff
@@ -5092,6 +5098,35 @@ public void renameTable(Database db, OlapTable table, TableRenameClause tableRen
         LOG.info("rename table[{}] to {}", tableName, newTableName);
     }
 
+    public void renameTable(Database db, Table table, TableRenameClause tableRenameClause) throws DdlException {
```
This method is just same as `renameTable` above?

HappenLee replied:
yes, but the param is different. one is the `olaptable`, this is `table`

HappenLee committed:
fix some mistake

HappenLee committed:
use GJSON to do Serialize

morningman reviewed (approved):
LGTM
